Defaults content_type to application/octet-stream in blob_record.js

weilandia opened the pull request:
### Summary

Defaults content_type to application/octet-stream in blob_record.js in order to resolve https://github.com/rails/rails/issues/38123.

@cushingw recommended this solution here https://github.com/rails/rails/issues/36514#issuecomment-503221829.

Closes https://github.com/rails/rails/issues/38123

weilandia commented:
Another solution could be to be more explicit and only change the content-type when the filename has .xls

javan commented on activestorage/app/javascript/activestorage/blob_record.js:
```diff
@@ -3,10 +3,11 @@ import { getMetaValue } from "./helpers"
     this.attributes = {
       filename: file.name,
-      content_type: file.type,
+      content_type: content_type,
```
I *think* we'd be better off not sending a `content_type` at all when `file.type` is blank and letting Active Storage extract one for us: 
https://github.com/rails/rails/blob/be40e8e5afc6bcf6092392741eec7dcd02edd14a/activestorage/app/models/active_storage/blob.rb#L208
Mind giving that a try?

weilandia replied:
The problem is the Content-Type we get back in the response from the the direct_uploads controller, which is then sent to `/rails/active_storage/disk`.

I tried the following (separately):

```
# blob_record.js:7
this.attributes = {
    filename: file.name,
    byte_size: file.size,
    checksum: checksum
  }

if (file.type) {
  this.attributes.content_type = file.type;
}

# resulted in Mime::Type::InvalidMimeType: "null" is not a valid MIME type
```

```
# blob_upload.js:11

for (const key in headers) {
  this.xhr.setRequestHeader(key, headers[key] || null)
}

# resulted in Mime::Type::InvalidMimeType: "null" is not a valid MIME type
```

```
# blob_upload.js:11

for (const key in headers) {
  if (!headers[key) continue;
  this.xhr.setRequestHeader(key, headers[key])
}

# sends Content-Type over as html and results in 422
```

@javan

javan commented on activestorage/app/javascript/activestorage/blob_record.js:
```diff
@@ -3,10 +3,11 @@ import { getMetaValue } from "./helpers"
     this.attributes = {
       filename: file.name,
-      content_type: file.type,
+      content_type: content_type,
```
Let's shorten and inline the default: `content_type: file.type || "application/octet-stream",`

weilandia committed:
Defaults content_type to application/octet-stream

Defaults content_type to application/octet-stream in blob_record.js

[direct_upload_xls_in_chrome]

javan commented:
Thanks!

weilandia commented:
@javan I noticed https://contributors.rubyonrails.org wrongly credited this commit to the branch name because it was in square brackets in the commit.  Very much not important, but wondering if there's a way to manually overwrite?  Thanks

javan commented:
Wups! Opened https://github.com/rails/rails-contributors/pull/146 to fix that up.

javan commented:
All better now: https://contributors.rubyonrails.org/contributors/nick-weiland/commits

weilandia commented:
Thanks @javan!

carlaraya commented:
It seems like the npm package (https://www.npmjs.com/package/@rails/activestorage) hasn't reflected the changes in this commit yet. The PR conversation is most likely not the right place to ask, but how can I implement this commit without waiting for the npm package to update?

smartygus commented:
> It seems like the npm package (https://www.npmjs.com/package/@rails/activestorage) hasn't reflected the changes in this commit yet. The PR conversation is most likely not the right place to ask, but how can I implement this commit without waiting for the npm package to update?

@carlaraya I basically forked the Rails repo from the 6.0.2.1 release, cherry-picked the commit that made this fix, then extracted the `activestorage` subdirectory into [another git repo](https://github.com/smartygus/activestorage) to create a patched activestorage repo, which I could then use directly with yarn, because it supports adding from a git repo, but unfortunately not if the package is in a subdirectory of a repo (see [here](https://github.com/yarnpkg/yarn/issues/4725)).

So in my case, adding the patched package with yarn became: `yarn add https://github.com/smartygus/activestorage`

javan commented:
I just backported this in https://github.com/rails/rails/commit/2c19a6e12c241e6ac289343e248bb01a11018226 so it'll go out in the next release.

maurobender commented:
Great! Thank you so much!! I was just facing this issue after updating to rails 6.

When is the next update including this fix for the [npm package](https://www.npmjs.com/package/activestorage) planned to happen? The las update was 5 months ago.

smartygus commented:
@maurobender the package you linked to is no longer used for Rails 6 (for Rails 5 only it seems). You can find the updated package here: https://www.npmjs.com/package/@rails/activestorage

maurobender commented:
> @maurobender the package you linked to is no longer used for Rails 6 (for Rails 5 only it seems). You can find the updated package here: https://www.npmjs.com/package/@rails/activestorage

Thanks a lot @smartygus! Any chance of adding this to the guides (maybe [here](https://edgeguides.rubyonrails.org/active_storage_overview.html#direct-uploads))?

smartygus commented:
> Any chance of adding this to the guides (maybe [here](https://edgeguides.rubyonrails.org/active_storage_overview.html#direct-uploads))?

@maurobender do you mean something to explain that the npm packages have moved? This is mentioned in the guide for upgrading from Rails 5.2 to Rails 6, and I think that's probably the most appropriate place to document such changes: https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#all-npm-packages-have-been-moved-to-the-@rails-scope
As far as I can tell, all mentions of the ActiveStorage npm package in the guides have also been updated to use the @rails/activestorage scoped package name, so I think the guides seem okay in terms of pointing people to the correct package.

Or did you have something else in mind?

betesh commented on activestorage/app/assets/javascripts/activestorage.js:
```diff
@@ -550,7 +550,7 @@
       this.attributes = {
         filename: file.name,
-        content_type: file.type,
+        content_type: file.type || "application/octet-stream",
```
Does it makes sense to set `null: false` for content_type in the migration, since it should never be NULL and causes errors if it is NULL?

ghiculescu replied:
That sounds like a good idea. Would you like to make a PR?

betesh replied:
Done: #43055
